File: go/processor/extractor/error_event.go
// @docspec:module {
//   id: "docspec-go-extractor-error-event",
//   name: "Error & Event Extractor",
//   description: "Detects sentinel errors (ErrXxx), custom error types, event structs, and docspec:error/docspec:event comment tags to populate the errors and events sections.",
//   since: "3.0.0"
// }
package extractor

import (
	"go/ast"
	"strconv"
	"strings"
)

// ErrorEventExtractor detects error codes and event definitions in Go source code:
// @docspec:boundary "classpath-safe extraction"
//   - Custom error types (implementing the error interface)
//   - Sentinel errors (var ErrXxx = errors.New(...))
//   - Error constants with codes
//   - Event structs and event publishing patterns
//   - Comment-based docspec tags (//docspec:error, //docspec:event)
//
// This is the Go equivalent of the Java error/event extraction that detects
// @DocError and @DocEvent annotations and populates the errors and events
// sections of the DocSpec model.
type ErrorEventExtractor struct{}

// @docspec:deterministic
func (e *ErrorEventExtractor) Name() string {
	return "error-event"
}

// IsAvailable always returns true since error/event detection works via
// naming conventions and doc comment tags.
//
// @docspec:deterministic
func (e *ErrorEventExtractor) IsAvailable(ctx *ProcessorContext) bool {
	return true
}

// Extract scans files for error and event patterns.
//
// @docspec:method { since: "3.0.0" }
// @docspec:intentional "Detects errors via sentinel variables, error types, and docspec tags; detects events via naming conventions and docspec tags"
func (e *ErrorEventExtractor) Extract(ctx *ProcessorContext) {
	for _, file := range ctx.Files {
		pkg := file.Name.Name
		e.extractErrors(file, pkg, ctx)
		e.extractEvents(file, pkg, ctx)
	}
}

// extractErrors finds error declarations, error types, and docspec:error tags.
//
// @docspec:intentional "Discovers sentinel errors, error types, and docspec:error comment tags"
func (e *ErrorEventExtractor) extractErrors(file *ast.File, pkg string, ctx *ProcessorContext) {
	for _, decl := range file.Decls {
		switch d := decl.(type) {
		case *ast.GenDecl:
			for _, spec := range d.Specs {
				// Check for sentinel errors: var ErrXxx = errors.New("...")
				if vs, ok := spec.(*ast.ValueSpec); ok {
					for i, name := range vs.Names {
						if !strings.HasPrefix(name.Name, "Err") || !name.IsExported() {
							continue
						}

						entry := ErrorEntry{
							Code:      name.Name,
							Exception: pkg + "." + name.Name,
						}

						// Extract description from the errors.New or fmt.Errorf argument
						if i < len(vs.Values) {
							entry.Description = extractErrorMessage(vs.Values[i])
						}

						// Check for docspec:error comment tag
						if d.Doc != nil {
							enrichErrorFromDoc(d.Doc, &entry)
						}
						if vs.Doc != nil {
							enrichErrorFromDoc(vs.Doc, &entry)
						}

						ctx.Errors = append(ctx.Errors, entry)
					}
				}

				// Check for error types (structs that may implement error interface)
				if ts, ok := spec.(*ast.TypeSpec); ok {
					if !ts.Name.IsExported() {
						continue
					}
					name := ts.Name.Name
					if !strings.HasSuffix(name, "Error") && !strings.HasSuffix(name, "Err") {
						continue
					}

					entry := ErrorEntry{
						Code:      name,
						Exception: pkg + "." + name,
					}

					// Try to get description from doc comment
					if d.Doc != nil {
						desc := docText(d.Doc)
						if desc != "" {
							entry.Description = desc
						}
						enrichErrorFromDoc(d.Doc, &entry)
					}

					ctx.Errors = append(ctx.Errors, entry)
				}
			}

		case *ast.FuncDecl:
			// Check function-level docspec:error tags
			if d.Doc == nil {
				continue
			}
			for _, c := range d.Doc.List {
				text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
				if !strings.HasPrefix(text, "docspec:error") {
					continue
				}
				// Format: //docspec:error CODE [httpStatus=NNN] [description]
				parts := strings.Fields(text)
				if len(parts) < 2 {
					continue
				}

				entry := ErrorEntry{
					Code: parts[1],
				}

				thrownBy := pkg + "." + d.Name.Name
				if d.Recv != nil && len(d.Recv.List) > 0 {
					thrownBy = pkg + "." + recvTypeName(d.Recv.List[0].Type) + "." + d.Name.Name
				}
				entry.ThrownBy = []string{thrownBy}

				for _, part := range parts[2:] {
					if strings.HasPrefix(part, "httpStatus=") {
						val := strings.TrimPrefix(part, "httpStatus=")
						if n, err := strconv.Atoi(val); err == nil {
							entry.HTTPStatus = n
						}
					}
				}

				// Remaining parts after flag-like params form the description
				desc := extractDescriptionFromParts(parts[2:])
				if desc != "" {
					entry.Description = desc
				}

				ctx.Errors = append(ctx.Errors, entry)
			}
		}
	}
}

// extractEvents finds event structs and docspec:event tags.
//
// @docspec:intentional "Discovers event structs by naming convention and docspec:event comment tags"
func (e *ErrorEventExtractor) extractEvents(file *ast.File, pkg string, ctx *ProcessorContext) {
	for _, decl := range file.Decls {
		switch d := decl.(type) {
		case *ast.GenDecl:
			for _, spec := range d.Specs {
				ts, ok := spec.(*ast.TypeSpec)
				if !ok || !ts.Name.IsExported() {
					continue
				}

				name := ts.Name.Name

				// Check for docspec:event comment tag
				if d.Doc != nil {
					for _, c := range d.Doc.List {
						text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
						if strings.HasPrefix(text, "docspec:event") {
							event := parseEventTag(text, name)
							ctx.Events = append(ctx.Events, event)
						}
					}
				}

				// Detect event structs by naming convention
				if !strings.HasSuffix(name, "Event") && !strings.HasSuffix(name, "Message") &&
					!strings.HasSuffix(name, "Notification") {
					continue
				}
				if _, ok := ts.Type.(*ast.StructType); !ok {
					continue
				}

				// Check if already added via docspec tag
				alreadyAdded := false
				for _, ev := range ctx.Events {
					if ev.Name == name {
						alreadyAdded = true
						break
					}
				}
				if alreadyAdded {
					continue
				}

				event := EventEntry{
					Name: name,
				}

				// Get description from doc comment
				if d.Doc != nil {
					desc := docText(d.Doc)
					if desc != "" {
						event.Description = desc
					}
				}

				ctx.Events = append(ctx.Events, event)
			}

		case *ast.FuncDecl:
			// Check function-level docspec:event tags
			if d.Doc == nil {
				continue
			}
			for _, c := range d.Doc.List {
				text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
				if strings.HasPrefix(text, "docspec:event") {
					event := parseEventTag(text, "")
					ctx.Events = append(ctx.Events, event)
				}
			}
		}
	}
}

// extractErrorMessage tries to extract the error message from an expression like
// errors.New("message") or fmt.Errorf("message", ...).
//
// @docspec:deterministic
func extractErrorMessage(expr ast.Expr) string {
	call, ok := expr.(*ast.CallExpr)
	if !ok {
		return ""
	}

	// Check for errors.New or fmt.Errorf
	sel, ok := call.Fun.(*ast.SelectorExpr)
	if !ok {
		return ""
	}

	ident, ok := sel.X.(*ast.Ident)
	if !ok {
		return ""
	}

	if (ident.Name == "errors" && sel.Sel.Name == "New") ||
		(ident.Name == "fmt" && sel.Sel.Name == "Errorf") {
		if len(call.Args) > 0 {
			if lit, ok := call.Args[0].(*ast.BasicLit); ok {
				return strings.Trim(lit.Value, `"`)
			}
		}
	}

	return ""
}

// enrichErrorFromDoc parses docspec error attributes from a comment group.
//
// @docspec:intentional "Enriches an ErrorEntry with httpStatus, resolution, and cause from docspec comment tags"
func enrichErrorFromDoc(doc *ast.CommentGroup, entry *ErrorEntry) {
	for _, c := range doc.List {
		text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))

		if strings.HasPrefix(text, "docspec:error") {
			parts := strings.Fields(text)
			for _, part := range parts[1:] {
				if strings.HasPrefix(part, "httpStatus=") {
					val := strings.TrimPrefix(part, "httpStatus=")
					if n, err := strconv.Atoi(val); err == nil {
						entry.HTTPStatus = n
					}
				}
				if strings.HasPrefix(part, "resolution=") {
					entry.Resolution = strings.TrimPrefix(part, "resolution=")
				}
			}
		}

		if strings.HasPrefix(text, "docspec:cause") {
			cause := strings.TrimPrefix(text, "docspec:cause ")
			cause = strings.TrimSpace(cause)
			if cause != "" {
				entry.Causes = append(entry.Causes, cause)
			}
		}
	}
}

// parseEventTag parses a //docspec:event tag.
// Format: //docspec:event NAME [channel=CHANNEL] [trigger=TRIGGER] [guarantee=GUARANTEE] [DESCRIPTION]
//
// @docspec:deterministic
func parseEventTag(text, defaultName string) EventEntry {
	parts := strings.Fields(text)
	event := EventEntry{}

	if len(parts) >= 2 {
		event.Name = parts[1]
	} else if defaultName != "" {
		event.Name = defaultName
	}

	for _, part := range parts[2:] {
		if strings.HasPrefix(part, "channel=") {
			event.Channel = strings.TrimPrefix(part, "channel=")
		} else if strings.HasPrefix(part, "trigger=") {
			event.Trigger = strings.TrimPrefix(part, "trigger=")
		} else if strings.HasPrefix(part, "guarantee=") {
			event.DeliveryGuarantee = strings.TrimPrefix(part, "guarantee=")
		}
	}

	desc := extractDescriptionFromParts(parts[2:])
	if desc != "" {
		event.Description = desc
	}

	return event
}

// docText extracts plain text from a comment group, stripping comment markers
// and docspec directive lines.
//
// @docspec:deterministic
func docText(doc *ast.CommentGroup) string {
	if doc == nil {
		return ""
	}
	var lines []string
	for _, c := range doc.List {
		text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
		if strings.HasPrefix(text, "docspec:") {
			continue
		}
		if text != "" {
			lines = append(lines, text)
		}
	}
	return strings.Join(lines, " ")
}

// extractDescriptionFromParts builds a description from parts that are not
// flag-like (key=value) parameters.
//
// @docspec:deterministic
func extractDescriptionFromParts(parts []string) string {
	var descParts []string
	for _, part := range parts {
		if strings.Contains(part, "=") {
			continue
		}
		descParts = append(descParts, part)
	}
	return strings.Join(descParts, " ")
}

File: rust/processor/src/extractor/error_event.rs
//! Detect error types and event patterns in Rust projects.
//!
//! @docspec:module {
//!   id: "docspec-rust-extractor-error-event",
//!   name: "Error & Event Extractor",
//!   description: "Detects error enums (thiserror, naming conventions, doc_error attributes) and event types (naming conventions, doc_event attributes) to populate the errors and events sections of the DocSpec model.",
//!   since: "3.0.0"
//! }
//!
//! Looks for:
//! - `thiserror` derive macros (`#[derive(Error)]`)
//! - Custom error enums implementing `std::error::Error`
//! - `Result<T, E>` return types to extract error type names
//! - Event structs / enums (names ending in Event, Message, Command, etc.)
//! - `#[doc_error]` and `#[doc_event]` custom attributes

use serde_json::Value;
use quote::ToTokens;
use crate::scanner::FileInfo;
use super::{DocSpecExtractor, ProcessorContext};

/// Suffixes that indicate an event type.
const EVENT_SUFFIXES: &[&str] = &[
    "Event",
    "Message",
    "Command",
    "Notification",
    "Signal",
    "Msg",
];

/// Suffixes that indicate an error type.
const ERROR_SUFFIXES: &[&str] = &[
    "Error",
    "Err",
    "Failure",
    "Exception",
];

/// @docspec:boundary "Error type and event pattern detection across enums, structs, and custom attributes"
pub struct ErrorEventExtractor;

impl DocSpecExtractor for ErrorEventExtractor {
    /// @docspec:deterministic
    fn is_available(&self, _ctx: &ProcessorContext) -> bool {
        // Error/event extraction is always available; it works with standard
        // Rust patterns and optional custom attributes.
        true
    }

    /// @docspec:deterministic
    fn extractor_name(&self) -> &'static str {
        "error-event"
    }

    /// @docspec:intentional "Detect error enums/structs and event types via derive macros, naming conventions, and custom attributes"
    fn extract(&self, files: &[FileInfo], ctx: &mut ProcessorContext) {
        let mut errors: Vec<Value> = Vec::new();
        let mut events: Vec<Value> = Vec::new();

        for file_info in files {
            let syntax = match syn::parse_file(&file_info.source) {
                Ok(s) => s,
                Err(_) => continue,
            };

            for item in &syntax.items {
                match item {
                    syn::Item::Enum(e) => {
                        let name = e.ident.to_string();
                        let qualified = format!("{}::{}", file_info.module, name);
                        let doc = crate::reader::doc_comment_reader::read_doc_comment(&e.attrs);

                        // Check if it's an error enum
                        if is_error_type(&e.attrs, &name) {
                            let variants: Vec<Value> = e.variants.iter().map(|v| {
                                let variant_name = v.ident.to_string();
                                let variant_doc = crate::reader::doc_comment_reader::read_doc_comment(&v.attrs);
                                let mut entry = serde_json::json!({
                                    "name": variant_name,
                                });
                                if let Some(d) = variant_doc {
                                    entry.as_object_mut().unwrap()
                                        .insert("description".to_string(), Value::String(d));
                                }
                                // Try to extract HTTP status code from #[error] attribute message
                                if let Some(status) = extract_error_status(&v.attrs) {
                                    entry.as_object_mut().unwrap()
                                        .insert("httpStatus".to_string(), Value::Number(status.into()));
                                }
                                entry
                            }).collect();

                            let mut error_entry = serde_json::json!({
                                "type": qualified,
                                "variants": variants,
                            });
                            if let Some(ref d) = doc {
                                error_entry.as_object_mut().unwrap()
                                    .insert("description".to_string(), Value::String(d.clone()));
                            }

                            // Check for #[doc_error] attributes
                            if let Some(attrs) = extract_doc_error_attrs(&e.attrs) {
                                let map = error_entry.as_object_mut().unwrap();
                                for (k, v) in attrs {
                                    map.insert(k, Value::String(v));
                                }
                            }

                            errors.push(error_entry);
                        }

                        // Check if it's an event enum
                        if is_event_type(&e.attrs, &name) {
                            let variants: Vec<String> = e.variants.iter()
                                .map(|v| v.ident.to_string())
                                .collect();
                            let mut event_entry = serde_json::json!({
                                "type": qualified,
                                "variants": variants,
                            });
                            if let Some(ref d) = doc {
                                event_entry.as_object_mut().unwrap()
                                    .insert("description".to_string(), Value::String(d.clone()));
                            }
                            if let Some(attrs) = extract_doc_event_attrs(&e.attrs) {
                                let map = event_entry.as_object_mut().unwrap();
                                for (k, v) in attrs {
                                    map.insert(k, Value::String(v));
                                }
                            }
                            events.push(event_entry);
                        }
                    }
                    syn::Item::Struct(s) => {
                        let name = s.ident.to_string();
                        let qualified = format!("{}::{}", file_info.module, name);
                        let doc = crate::reader::doc_comment_reader::read_doc_comment(&s.attrs);

                        // Check if struct has #[doc_error]
                        if has_attr(&s.attrs, "doc_error") || is_name_match(&name, ERROR_SUFFIXES) {
                            let fields: Vec<String> = s.fields.iter()
                                .filter_map(|f| f.ident.as_ref().map(|i| i.to_string()))
                                .collect();
                            let mut error_entry = serde_json::json!({
                                "type": qualified,
                                "fields": fields,
                            });
                            if let Some(ref d) = doc {
                                error_entry.as_object_mut().unwrap()
                                    .insert("description".to_string(), Value::String(d.clone()));
                            }
                            errors.push(error_entry);
                        }

                        // Check if struct has #[doc_event] or name matches event pattern
                        if has_attr(&s.attrs, "doc_event") || is_name_match(&name, EVENT_SUFFIXES) {
                            let fields: Vec<String> = s.fields.iter()
                                .filter_map(|f| f.ident.as_ref().map(|i| i.to_string()))
                                .collect();
                            let mut event_entry = serde_json::json!({
                                "type": qualified,
                                "fields": fields,
                            });
                            if let Some(ref d) = doc {
                                event_entry.as_object_mut().unwrap()
                                    .insert("description".to_string(), Value::String(d.clone()));
                            }
                            if let Some(attrs) = extract_doc_event_attrs(&s.attrs) {
                                let map = event_entry.as_object_mut().unwrap();
                                for (k, v) in attrs {
                                    map.insert(k, Value::String(v));
                                }
                            }
                            events.push(event_entry);
                        }
                    }
                    _ => {}
                }
            }
        }

        if !errors.is_empty() {
            ctx.errors = errors;
        }
        if !events.is_empty() {
            ctx.events = events;
        }
    }
}

/// Check if a type is an error type by derive macros or naming convention.
///
/// @docspec:deterministic
fn is_error_type(attrs: &[syn::Attribute], name: &str) -> bool {
    // Check for #[derive(Error)] (thiserror)
    if has_derive(attrs, "Error") {
        return true;
    }
    // Check for #[doc_error]
    if has_attr(attrs, "doc_error") {
        return true;
    }
    // Check naming convention
    is_name_match(name, ERROR_SUFFIXES)
}

/// Check if a type is an event type by naming convention or attribute.
///
/// @docspec:deterministic
fn is_event_type(attrs: &[syn::Attribute], name: &str) -> bool {
    if has_attr(attrs, "doc_event") {
        return true;
    }
    is_name_match(name, EVENT_SUFFIXES)
}

/// Check if a name ends with any of the given suffixes.
///
/// @docspec:deterministic
fn is_name_match(name: &str, suffixes: &[&str]) -> bool {
    suffixes.iter().any(|suffix| name.ends_with(suffix))
}

/// Check if a derive attribute contains the given macro name.
///
/// @docspec:deterministic
fn has_derive(attrs: &[syn::Attribute], derive_name: &str) -> bool {
    for attr in attrs {
        if attr.path().is_ident("derive") {
            let tokens = attr.meta.to_token_stream().to_string();
            if tokens.contains(derive_name) {
                return true;
            }
        }
    }
    false
}

/// Check if an attribute with the given name is present.
///
/// @docspec:deterministic
fn has_attr(attrs: &[syn::Attribute], name: &str) -> bool {
    attrs.iter().any(|a| {
        a.path().get_ident().map(|i| i == name).unwrap_or(false)
    })
}

/// Try to extract an HTTP status code from `#[error(status = 404)]` or similar patterns.
///
/// @docspec:deterministic
fn extract_error_status(attrs: &[syn::Attribute]) -> Option<u16> {
    for attr in attrs {
        let tokens = attr.meta.to_token_stream().to_string();
        // Look for status = NNN pattern
        if let Some(pos) = tokens.find("status") {
            let rest = &tokens[pos + 6..];
            let rest = rest.trim_start().trim_start_matches('=').trim_start();
            let num_str: String = rest.chars().take_while(|c| c.is_ascii_digit()).collect();
            if let Ok(status) = num_str.parse::<u16>() {
                if (100..600).contains(&status) {
                    return Some(status);
                }
            }
        }
    }
    None
}

/// Extract key-value pairs from `#[doc_error(code = "E001", severity = "critical")]`.
///
/// @docspec:deterministic
fn extract_doc_error_attrs(attrs: &[syn::Attribute]) -> Option<Vec<(String, String)>> {
    for attr in attrs {
        if let Some(ident) = attr.path().get_ident() {
            if ident == "doc_error" {
                let tokens = attr.meta.to_token_stream().to_string();
                let pairs = extract_all_kvs(&tokens);
                if !pairs.is_empty() {
                    return Some(pairs);
                }
            }
        }
    }
    None
}

/// Extract key-value pairs from `#[doc_event(channel = "orders", ...)`.
///
/// @docspec:deterministic
fn extract_doc_event_attrs(attrs: &[syn::Attribute]) -> Option<Vec<(String, String)>> {
    for attr in attrs {
        if let Some(ident) = attr.path().get_ident() {
            if ident == "doc_event" {
                let tokens = attr.meta.to_token_stream().to_string();
                let pairs = extract_all_kvs(&tokens);
                if !pairs.is_empty() {
                    return Some(pairs);
                }
            }
        }
    }
    None
}

/// Extract all `key = "value"` pairs from a token string.
///
/// @docspec:deterministic
fn extract_all_kvs(tokens: &str) -> Vec<(String, String)> {
    let mut pairs = Vec::new();
    // Find content inside outermost parentheses
    let inner = if let Some(start) = tokens.find('(') {
        if let Some(end) = tokens.rfind(')') {
            &tokens[start + 1..end]
        } else {
            return pairs;
        }
    } else {
        return pairs;
    };

    // Split by comma and parse each key = "value"
    for part in inner.split(',') {
        let part = part.trim();
        if let Some(eq_pos) = part.find('=') {
            let key = part[..eq_pos].trim().to_string();
            let val = part[eq_pos + 1..].trim();
            // Strip surrounding quotes
            let val = val.trim_matches('"').to_string();
            if !key.is_empty() && !val.is_empty() {
                pairs.push((key, val));
            }
        }
    }
    pairs
}
